frontend/apps/reader/modules/readercoptlistener.lua

```diff
@@ -54,6 +61,46 @@ function ReaderCoptListener:setAndSave(setting, property, value)
     UIManager:broadcastEvent(Event:new("SetStatusLine", self.document.configurable.status_line, true))
 end
 
+function ReaderCoptListener:UpdateHeader()
+    if self.view.view_mode == "page" then
+        if self.ui.document.provider == "crengine" then -- can that "if" be dropped
+            self.ui.document._callCacheSet("current_buffer_tag", nil)
+        end
+        self.ui:handleEvent(Event:new("RedrawCurrentView", self.current_page))
```

Comment:
Yes, it can be dropped, ReaderCoptListener being loaded mean crengine is being used.
`self.ui.document._callCacheSet("current_buffer_tag", nil)` is ugly and some internal function.
edec69ac8 added a no-op `resetCallCache()` - but that's not what you want.
You want a similarly set up function that could be called trashBufferCache() and would just have:
`elseif name == "resetCallCache" then add_buffer_trash = true`.

After that, I think you don't even need to have RedrawCurrentView called: simply calling self.ui.rolling:updateBatteryState(), and then a simple setDirty should be enough for UIManager to redraw ReaderUI, asking crengine to draw it again, hopefully with the updated time and battery.
For the setDirty region, we have the height with `CreDocument:getHeaderHeight()`.May be this would work:
```lua
            UIManager:setDirty(self.view.dialog, function()
                return "ui", Geom:new{ w = Screen:getWidth(), h = self.ui.document:getHeaderHeight()},
            end)
```
(I let @NiLuJe tell us what's the right way to do it.)

Reply:
The header is not refreshed when I change `updateHeader` to:
```
function ReaderCoptListener:updateHeader()
    if self.view.view_mode == "page" then
        require("logger").dbg("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx updateHeader")
--        self.ui.document._callCacheSet("current_buffer_tag", nil)
--        self.ui:handleEvent(Event:new("RedrawCurrentView", self.current_page))
        self.ui.rolling:updateBatteryState()
        UIManager:setDirty(self.view.dialog, function()
            return "ui", Geom:new{ w = Device.screen:getWidth(), h = self.ui.document:getHeaderHeight()}
        end)
    end
end
```

the log shows:
```
01/21/21-21:24:01 DEBUG xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx updateHeader
01/21/21-21:24:01 DEBUG update battery state
01/21/21-21:24:01 DEBUG CreDocument: set battery state 88
01/21/21-21:24:01 DEBUG setDirty via a func from widget ReaderUI
01/21/21-21:24:01 DEBUG painting widget: ReaderUI
01/21/21-21:24:01 DEBUG callCache: drawCurrentViewByPage setting tag
01/21/21-21:24:01 DEBUG callCache: ---------- drawCurrentView: light draw
01/21/21-21:24:01 DEBUG _refresh: Enqueued ui update for region 0 706 540 14 
01/21/21-21:24:01 DEBUG callCache: getHeaderHeight cache miss: getHeaderHeight
01/21/21-21:24:01 DEBUG _refresh: Enqueued ui update for region 0 0 540 31 
01/21/21-21:24:01 DEBUG refresh on physical rectangle 0 706 540 14
01/21/21-21:24:01 DEBUG refresh on physical rectangle 0 0 540 31
```

(I will wait for @NiLuJe for input.)

Reply:
drawCurrentView: light draw = buffer not trashed.
You need to add resetBufferCache into credocument.lua as I outlined above, and call it here
```
self.ui.rolling:updateBatteryState()
self.ui.document:resetBufferCache()
UIManager..setDirty...
```

Reply:
I think I got it now. -> I will do that tomorrow.

Reply:
Doesn't even need to be a function if you already know the exact dimension of the refresh ;).

(The "pass a function to setDirty" trick is necessary for stuff that uses a widget's internal `dimen` field as the region, as those as (*generally*) only computed during the widget's `paintTo`).

e.g.,

https://github.com/koreader/koreader/blob/9f4e7456cfc105383cead062eb9cb51d7463eab5/frontend/ui/uimanager.lua#L636